pandas/core/groupby/ops.py

```diff
@@ -452,6 +453,16 @@ def wrapper(*args, **kwargs):
     def _cython_operation(self, kind, values, how, axis, min_count=-1, **kwargs):
         assert kind in ["transform", "aggregate"]
 
+        if is_datetime64tz_dtype(values):
```

Comment:
you shouldn't do this here and rather handle at a higher level as datetime64 is caught below for some operations.

Reply:
later in this function it checks for datetime64 to exclude add/prod/cumsum/cumprod, but there is nowhere else in this module where we handle datetime64tz AFAICT

Reply:
my point is that this is not the right place to handle this. if you want to transform the input then let it pass thru or raise a NotImplementedError then that is fine, but the way you are writing it is is *very* diffferent.

Reply:
all `aggregate` and `transform` calls go through here, so this is the only place AFAICT that we can do this without having to duplicate it for every calling function.  If there is a better place, can you be more specific?

Reply:
missing my point. I have no problem handling this in this function. What you wrote returns early, bypassing *all* of the logic that handles all of the other dtypes. So it needs to be integrated there OR handled at a higher level on a NotImplementedError raise.

This function becomes *even* harder to understand with this addition. In theory you could also split this function up more logically. There are a number of solutions.

Reply:
edited to do what I think you're suggesting.

the method could use a good refactor, agreed.  separate pass.